[AArch64][MachineCombiner] Combine sequences of gather patterns

jcohen-apple opened the pull request:
Reland of #142941

Squashed with fixes for #150004, #149585

Notably added a check to conservatively apply the optimization, only if there is no chance of an aliasing instruction between the LD1 instructions we combine. If there are any function calls or stores - bail out.

jcohen-apple committed:
[AArch64][MachineCombiner] Combine sequences of gather patterns

jcohen-apple committed:
Use `mayAlias` on all memory and call instrs in the interval between FirstLoad and LastLoad

jcohen-apple commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7412,11 +7413,347 @@ static bool getMiscPatterns(MachineInstr &Root,
 
+/// Check if there are any stores or calls between two instructions in the same
+/// basic block.
+static bool hasInterveningStoreOrCall(const MachineInstr *First,
```
Sure. I was reluctant to add it initially because I didn't want to add AliasAnalysis as a dependency of MachineCombiner just for this pattern, but it seems to catch all the problematic patterns without it. Updated with an additional commit.

jcohen-apple committed:
Remove inefficient sort, add a parameter to limit the number of instructions checked.

jcohen-apple commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7412,11 +7413,347 @@ static bool getMiscPatterns(MachineInstr &Root,
+  SmallVector<const MachineInstr *, 16> SortedLoads(LoadInstrs.begin(),
+                                                    LoadInstrs.end());
+  llvm::sort(SortedLoads, [](const MachineInstr *A, const MachineInstr *B) {
+    if (A->getParent() != B->getParent()) {
```
Good catch. I added this to get something working but forgot to get back to it.
I added a parameter to limit the number of instructions searched, and implement an iteration going up the basic block starting from Root, and exiting once all loads have been encountered or we have hit the limit.

jcohen-apple committed:
Rename functions for CR

jcohen-apple commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7412,11 +7413,347 @@ static bool getMiscPatterns(MachineInstr &Root,
+/// sequential NEON lane loads into parallel vector loads that can execute
+/// concurrently.
+static void
+generateGatherPattern(MachineInstr &Root,
```
Done

jcohen-apple commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7412,11 +7413,347 @@ static bool getMiscPatterns(MachineInstr &Root,
+///   - 32-bit elements (LD1i32, 4 lanes total)
+///   - 16-bit elements (LD1i16, 8 lanes total)
+///   - 8-bit elements (LD1i8, 16 lanes total)
+static bool getGatherPattern(MachineInstr &Root,
```
Done

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7412,11 +7419,335 @@ static bool getMiscPatterns(MachineInstr &Root,
+                     AliasAnalysis *AA) {
+  for (const MachineInstr *MIb : MemInstrs) {
+    if (MIa.mayAlias(AA, *MIb, /*UseTBAA*/ false)) {
+      MIb->dump();
```
Remove dump

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7412,11 +7419,335 @@ static bool getMiscPatterns(MachineInstr &Root,
+  auto *CurrInstr = MRI.getUniqueVRegDef(Root.getOperand(1).getReg());
+  auto Range = llvm::seq<unsigned>(1, NumLanes - 1);
+  SmallSet<unsigned, 16> RemainingLanes(Range.begin(), Range.end());
+  SmallVector<const MachineInstr *, 16> LoadInstrs = {};
```
I think this can remove the ` = {}`

davemgreen commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -7412,11 +7419,335 @@ static bool getMiscPatterns(MachineInstr &Root,
+    // Check for potential aliasing with any of the load instructions to
+    // optimize.
+    if ((CurrInstr.mayLoadOrStore() || CurrInstr.isCall()) &&
+        mayAlias(CurrInstr, LoadInstrs, nullptr))
```
AA is always nullptr?
Could it use isLoadFoldBarrier?

davemgreen reviewed (commented):
Thanks - I had forgotten this could just go through the instructions in order, that sounds much better.

jcohen-apple committed:
Address CR comment, pt.2

Remove the `mayAlias` check, and only check if there are any instructions between the loads we want to optimize that could be a load fold barrier instead.

davemgreen reviewed (approved):
LGTM, thanks

jcohen-apple committed:
Merge branch 'main' into users/jcohen-apple/gather-opt-reland
